syncopy/tests/backend/test_connectivity.py

```diff
@@ -0,0 +1,324 @@
+                   '-o', label='original CSD', ms=3)
+    assert ax.plot(freqs, np.abs(CSDreg[:, chan, chan]),
+                   '-o', label='regularized CSD', ms=3)
+    assert ax.plot(freqs, np.abs(CSDfac[:, chan, chan]),
+                   '-o', label='factorized CSD', ms=3)
+    ax.set_xlim((f1 - 5, f2 + 5))
+
+
+def test_granger():
+
+    '''
+    Test the granger causality measure
+    with uni-directionally coupled AR(2)
+    processes akin to the source publication:
+
+    Dhamala, Mukeshwar, Govindan Rangarajan, and Mingzhou Ding. 
+       "Estimating Granger causality from Fourier and wavelet transforms 
+        of time series data." Physical review letters 100.1 (2008): 018701.
+    '''
+
+    fs = 200 # Hz
+    nSamples = 2500
+    nTrials = 50
+    
+    # both AR(2) processes have same parameters
+    # and yield a spectral peak at 40Hz
+    alpha1, alpha2 = 0.55, -0.8
+    coupling = 0.25
+
+    CSDav = np.zeros((nSamples // 2 + 1, 2, 2), dtype=np.complex64)
+    for _ in range(nTrials):
+
+        # -- simulate 2 AR(2) processes --
+        
+        sol = np.zeros((nSamples, 2))
+        # pick the 1st values at random
+        xs_ini = np.random.randn(2, 2)
+        sol[:2, :] = xs_ini
+        for i in range(1, nSamples):
+            sol[i, 1] = alpha1 * sol[i - 1, 1] + alpha2 * sol[i - 2, 1] 
+            sol[i, 1] += np.random.randn()
+            # X2 drives X1
+            sol[i, 0] = alpha1 * sol[i - 1, 0] + alpha2 * sol[i - 2, 0]
+            sol[i, 0] += sol[i - 1, 1] * coupling 
+            sol[i, 0] += np.random.randn()
+
+        # --- get CSD ---
+        bw = 5
+        NW = bw * nSamples / (2 * 1000)
+        Kmax = int(2 * NW - 1) # optimal number of tapers
+        CS2, freqs = stCR.cross_spectra_cF(sol, fs,
+                                           taper='dpss',
+                                           taper_opt={'Kmax' : Kmax, 'NW' : NW},
+                                           fullOutput=True)
+
+        CSD = CS2[0, ...]
+        CSDav += CSD
+            
+    CSDav /= nTrials
+    # with only 2 channels this CSD is well conditioned
+    assert np.linalg.cond(CSDav).max() < 1e2
+    H, Sigma, conv = wilson_sf(CSDav)
+    
+    G = granger(CSDav, H, Sigma)
+    assert G.shape == CSDav.shape
+
+    # check for directional causality at 40Hz
+    freq_idx = np.argmin(freqs < 40)
+    assert 39 < freqs[freq_idx] < 41
+
+    # check low to no causality for 1->2
+    assert G[freq_idx, 0, 1] < 0.1
+    # check high causality for 2->1
+    assert G[freq_idx, 1, 0] > 0.8
+    
+    fig, ax = ppl.subplots(figsize=(6, 4))
+    ax.set_xlabel('frequency (Hz)')
+    ax.set_ylabel(r'Granger causality(f)')
+    assert ax.plot(freqs, G[:, 0, 1], label=r'Granger $1\rightarrow2$')
+    assert ax.plot(freqs, G[:, 1, 0], label=r'Granger $2\rightarrow1$')
+    ax.legend()
+
+# --- Helper routines ---
+
+
+# noisy phase evolution -> phase diffusion
+def phase_evo(omega0, eps, fs=1000, N=1000):
```

Comment:
This test signal is quite cool. We might want to keep this in mind for other tests too

Reply:
Yap, quite happy with this general concept too: moderate to high power in similiar frequency bands but yet low phase coherence due to phase diffusion :grinning: